README.md
# react-native-image-filter-kit
[![npm version](https://badge.fury.io/js/react-native-image-filter-kit.svg)](https://badge.fury.io/js/react-native-image-filter-kit)
[![js-semistandard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg)](https://github.com/standard/standard)
[![Dependency Status](https://david-dm.org/iyegoroff/react-native-image-filter-kit.svg)](https://david-dm.org/iyegoroff/react-native-image-filter-kit)
[![devDependencies Status](https://david-dm.org/iyegoroff/react-native-image-filter-kit/dev-status.svg)](https://david-dm.org/iyegoroff/react-native-image-filter-kit?type=dev)
[![npm](https://img.shields.io/npm/l/express.svg)](https://www.npmjs.com/package/react-native-image-filter-kit)

Various image filters for iOS & Android.

## Status

- iOS & Android:
  - filter components work as combinable wrappers for standard `Image` and `ImageBackground` components
  - resulting images are being cached in memory
  - <strong>Minimum Android SDK version - 17</strong>
- React-Native:
  - with rn >= 0.57.1 use latest version

## Getting started

`$ npm install react-native-image-filter-kit --save`

### Mostly automatic installation

`$ react-native link react-native-image-filter-kit`

### Manual installation

[link](docs/manual_installation.md)

### Final step

Open `android/build.gradle` and change `minSdkVersion` to 17.

## Scope

The purpose of this module is to support most of the native image filters on each platform and to provide a common interface for these filters. If the filter exists only on one platform, then its counterpart will be implemented using `renderscript` on Android and `cikernel` on iOS. If you need only [color matrix](docs/color_matrix_filters.md) filters - better use a [lightweight predecessor](https://github.com/iyegoroff/react-native-color-matrix-image-filters) of this module.


## Example

```javascript
import { Image } from 'react-native';
import {
  SoftLightBlend,
  Emboss,
  Earlybird,
  Invert,
  RadialGradient
} from 'react-native-image-filter-kit';

const result = (
  <Earlybird
    image={
      <SoftLightBlend
        resizeCanvasTo={'dstImage'}
        dstResizeMode={'CONTAIN'}
        dstImage={
          <Emboss
            image={
              <Image
                style={{ width: 320, height: 320 }}
                source={require('./parrot.png')}
                resizeMode={'contain'}
              />
            }
          />
        }
        srcAnchor={{ x: 0.5, y: 1 }}
        srcPosition={{ x: 0.5, y: 1 }}
        srcImage={
          <Invert
            image={
              <RadialGradient
                colors={['rgba(0, 0, 255, 1)', '#00ff00', 'red']}
                stops={[0.25, 0.75, 1]}
                center={{ x: '50w', y: '100h' }}
              />
            }
          />
        }
      />
    }
  />
)
```

<table>
  <tr>
    <th>original image</th>
    <th>result</th>
  </tr>
  <tr>
    <th><img src="https://raw.githubusercontent.com/iyegoroff/react-native-image-filter-kit/master/img/parrot.png" align="left" width="300"></th>
    <th><img src="https://raw.githubusercontent.com/iyegoroff/react-native-image-filter-kit/master/img/earlybird.png" align="left" width="300"></th>
  </tr>
</table>
&nbsp;
<details>
  <summary>filter steps</summary>
  <table>
  <tr>
    <th>
      <table>
        <tr><th>original image</th></tr>
        <tr><th><img src="https://raw.githubusercontent.com/iyegoroff/react-native-image-filter-kit/master/img/parrot.png" align="left" width="170"></th></tr>
      </table>
    </th>
    <th>
      <table>
        <tr><th>Emboss</th></tr>
        <tr><th><img src="https://raw.githubusercontent.com/iyegoroff/react-native-image-filter-kit/master/img/emboss.png" align="left" width="170"></th></tr>
      </table>
    </th>
    <th rowspan="2">
      <table>
        <tr><th>SoftLightBlend</th></tr>
        <tr><th><img src="https://raw.githubusercontent.com/iyegoroff/react-native-image-filter-kit/master/img/soft_light_blend.png" align="left" width="170"></th></tr>
      </table>
    </th>
    <th rowspan="2">
      <table>
        <tr><th>Earlybird</th></tr>
        <tr><th><img src="https://raw.githubusercontent.com/iyegoroff/react-native-image-filter-kit/master/img/earlybird.png" align="left" width="170"></th></tr>
      </table>
    </th>
  </tr>
  <tr>
    <td>
      <table>
        <tr><th>RadialGradient</th></tr>
        <tr><th><img src="https://raw.githubusercontent.com/iyegoroff/react-native-image-filter-kit/master/img/radial_gradient.png" align="left" width="170"></th></tr>
      </table>
    </td>
    <td>
      <table>
        <tr><th>Invert</th></tr>
        <tr><th><img src="https://raw.githubusercontent.com/iyegoroff/react-native-image-filter-kit/master/img/invert.png" align="left" width="170"></th></tr>
      </table>
    </td>
  </tr>
</table>

</details>

## Reference

- [Types](docs/types.md)
- [Functions](docs/functions.md)
- [Color matrix filters](docs/color_matrix_filters.md)
- [Blur filters](docs/blur_filters.md)
- [Convolve matrix filters](docs/convolve_matrix_filters.md)
- [Generators](docs/generators.md)
- [Composition filters](docs/composition_filters.md)
- [Blend filters](docs/blend_filters.md)
- [CSSGram filters](docs/cssgram_filters.md)
- [Android-only filters](https://github.com/iyegoroff/react-native-image-filter-kit/blob/master/src/native-platform-filters/shapes.android.ts)
- [iOS-only filters](https://github.com/iyegoroff/react-native-image-filter-kit/blob/master/src/native-platform-filters/shapes.ios.ts)

## Caveats
- `blurRadius` Image prop will not work in conjunction with this library, instead of it just use [BoxBlur](docs/blur_filters.md#BoxBlur) filter
- When running on pre-Lollipop (SDK < 21) Android devices you may experience [TooManyBitmapsException](https://frescolib.org/javadoc/reference/com/facebook/imagepipeline/common/TooManyBitmapsException.html), which results in image is not being rendered (this can be logged with [onFilteringError](docs/types.md#ImageFilter) prop). It looks like this is a relatively rare case which arises on low-end devices when filtering wallpaper-sized images (like 1920 × 1080 pixels). The common workarounds are:
  - using smaller images
  - using [ColorMatrix](docs/color_matrix_filters.md#ColorMatrix) filter with [concatColorMatrices](docs/functions.md#concatColorMatrices) instead of wrapping the image with multiple color matrix based filters
  - adding `android:largeHeap="true"` to `android/app/src/main/AndroidManifest.xml`
  - replacing standard `MainReactPackage` with [alternative](android/src/main/java/iyegoroff/imagefilterkit/MainReactPackageWithFrescoCache.java
) one provided by this module:
    ```diff
      ...
    - import com.facebook.react.shell.MainReactPackage;
    + import iyegoroff.imagefilterkit.MainReactPackageWithFrescoCache;
      import com.facebook.soloader.SoLoader;

      import java.util.Arrays;
      import java.util.List;

      public class MainApplication extends Application implements ReactApplication {

    +   private class MainReactPackage extends MainReactPackageWithFrescoCache {}
      ...
    ```
    After this change `ImageFilter` will not throw `TooManyBitmapsException` immediately and will clear Fresco image caches, trim bitmap pool memory and try to filter the image again several times until succeed or reach the limit of retries, specified by [clearCachesMaxRetries](docs/types.md#ImageFilter) prop.
- Note to `react-native-asset` users - linking this module will break your `Copy Bundle Resources` build phase on iOS, because there is no way to run `react-native-asset` in [incremental/decremental mode](https://github.com/unimonkiez/react-native-asset/issues/10) currently. So you need to link manually [this module](docs/manual_installation.md) or your own resources without `react-native-asset` usage.

## Credits
- CSSGram filters are taken from [cssgram](https://github.com/una/cssgram) project by @una
- `EdgeDetection`, `Emboss` and `FuzzyGlass` filters are taken from [android-graphics-demo](https://github.com/chiuki/android-graphics-demo) project by @chiuki
- parrot [image](https://commons.wikimedia.org/wiki/File:Ara_macao_-flying_away-8a.jpg) by
  [Robert01](https://de.wikipedia.org/wiki/Benutzer:Robert01)


RELEASE_NOTES.md
[release notes](https://github.com/iyegoroff/react-native-image-filter-kit/releases)

Simple/App.js
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow
 */

import React, {Component, PureComponent} from 'react';
import {
  Platform,
  StyleSheet,
  Text,
  View,
  Image,
  ScrollView,
  TouchableWithoutFeedback,
  FlatList,
  Picker,
  Switch,
  Button,
  PickerIOS
} from 'react-native';
import {
  ImageFilter,
  RoundAsCircle,
  IterativeBoxBlur,
  PorterDuffXfermode,
  Color,
  concatColorMatrices,
  luminanceToAlpha,
  rgbaToRgb
} from 'react-native-image-filter-kit';
import matrices from 'rn-color-matrices';
import { ShapeRegistry } from '../dist/common/shape-registry';
import { ImageTransparentPlaceholder, ImagePlaceholder } from '../dist/common/image-placeholder';

const imageStyle = { width: 320, height: 320 }
const Pick = Platform.OS === 'ios' ? PickerIOS : Picker
const PickItem = Platform.OS === 'ios' ? PickerIOS.Item : Picker.Item

// Image.prefetch('http://www.maximumwall.com/wp-content/uploads/2017/01/wallpaper-image-nourriture-hd-13.jpg');
// Image.prefetch('https://media.ooreka.fr/public/image/plant/314/mainImage-source-11702050.jpg');

const degToRad = (deg) => Math.PI * deg / 180;
const background = 'rgb(255, 255, 255)';
const resizeMode = 'contain'
const atx = (
  <Image
    style={imageStyle}
    source={{ uri: 'https://una.im/CSSgram/img/atx.jpg' }}
    resizeMode={resizeMode}
  />
);

const bike = (
  <Image
    style={imageStyle}
    source={{ uri: 'https://una.im/CSSgram/img/bike.jpg' }}
    resizeMode={resizeMode}
  />
);

const cacti = (
  <Image
    style={imageStyle}
    source={{ uri: 'https://una.im/CSSgram/img/cacti.jpg' }}
    resizeMode={resizeMode}
  />
);

const tahoe = (
  <Image
    style={imageStyle}
    source={{ uri: 'https://una.im/CSSgram/img/tahoe.jpg' }}
    resizeMode={resizeMode}
  />
)

const dest = (
  <Image
    style={imageStyle}
    source={require('./dest.png')}
    resizeMode={resizeMode}
  />
);

const src = (
  <Image
    style={imageStyle}
    source={require('./src.png')}
    resizeMode={resizeMode}
  />
);

const parrot = (
  <Image
    style={imageStyle}
    source={require('./parrot.png')}
    resizeMode={resizeMode}
  />
);

const rose = (
  <Image
    style={imageStyle}
    source={require('./rose.png')}
    resizeMode={resizeMode}
  />
);

const flowers = (
  <Image
    style={imageStyle}
    source={{ uri: 'https://media.ooreka.fr/public/image/plant/314/mainImage-source-11702050.jpg' }}
    resizeMode={resizeMode}
  />
);

const pizza = (
  <Image
    style={imageStyle}
    source={{ uri: 'http://www.maximumwall.com/wp-content/uploads/2017/01/wallpaper-image-nourriture-hd-13.jpg' }}
    resizeMode={resizeMode}
  />
);

const miniCoast = (
  <Image
    style={{ width: 320, height: 100, resizeMode: 'stretch' }}
    source={{ uri: 'https://thisismyhappiness.com/wp-content/uploads/2014/05/big-sur-bixby.jpg' }}
  />
)

const random = (
  <Image
    style={imageStyle}
    source={{ uri: 'https://picsum.photos/200/?random?t=' + Date.now() }}
    resizeMode={resizeMode}
  />
);

const instructions = Platform.select({
  ios: 'Press Cmd+R to reload,\n' + 'Cmd+D or shake for dev menu',
  android:
    'Double tap R on your keyboard to reload,\n' +
    'Shake or press menu button for dev menu',
});

const Img = (props) => <Image {...props} />

class CSSGramItem extends PureComponent {
  static defaultProps = {
    filter: 'Normal',
    image: require('./parrot.png')
  }

  halfTimer = null;
  timer = null;

  state = { isFiltered: true, isHalf: false };

  componentWillUnmount() {
    clearTimeout(this.timer);
    clearTimeout(this.halfTimer);
  }

  render() {
    const { isFiltered } = this.state;

    return (
      <View style={styles.filter}>
        <TouchableWithoutFeedback onPress={this.pressed}>
          {isFiltered ? this.filteredImage() : this.image()}
        </TouchableWithoutFeedback>
        {isFiltered
          ? <View style={styles.textWrap}><Text style={styles.text}>{this.props.filter}</Text></View>
          : null}
      </View>
    );
  }

  image() {
    const { image: uri } = this.props;

    return (
      <Img
        style={this.state.isHalf ? styles.halfImage : styles.image}
        source={typeof uri === 'number' ? uri : { uri }}
        resizeMode={'contain'}
      />
    );
  }

  filteredImage() {
    return (
      <ImageFilter
        onFilteringError={this.handleError}
        onFilteringStart={this.handleStart}
        onFilteringFinish={this.handleFinish}
        config={{
          name: this.props.filter,
          image: this.image(),
          disableCache: false,
        }}
      />
    );
  }

  handleStart = () => {
    console.warn('START')
  }

  handleFinish = () => {
    console.warn('FINISH')
  }

  handleError = (error) => {
    console.warn(error.nativeEvent.message)
  }

  pressed = () => {
    const { isFiltered } = this.state;

    // if (!isFiltered) {
    //   this.halfTimer = setTimeout(() => this.setState({ isHalf: true }), 1000);
    //   this.timer = setTimeout(() => this.setState({ isHalf: false }), 2000);
    // }
    this.setState({ isFiltered: !isFiltered });
  }
};

type Props = {};
export default class App extends Component<Props> {
  state = {
    t: Date.now(),
    blends: [
      'Modulate',
      'Multiply',
      'Difference',
      'Color',
      'Exclusion',
      'Luminosity',
      'HardLight',
      'Darken',
      'Saturation',
      'SoftLight',
      'Lighten',
      'Plus',
      'ColorBurn',
      'ColorDodge',
      'Hue',
      'Overlay',
      'Screen'
    ],
    selectedBlend: 'Lighten',
    blendImages: [
      { name: 'parrot', item: parrot },
      { name: 'tahoe', item: tahoe },
      { name: 'atx', item: atx },
      { name: 'cacti', item: cacti },
      { name: 'bike', item: bike },
      { name: 'dest', item: dest },
      { name: 'src', item: src },
      { name: 'pizza', item: pizza },
    ],
    firstSelectedImage: 'parrot',
    secondSelectedImage: 'tahoe',
    showList: false,
    selectedFilter: 'Normal_0',
    selectedImage: 'Parrot',
    images: [
      { name: 'Parrot', uri: require('./parrot.png') },
      { name: 'Pizza', uri: 'http://www.maximumwall.com/wp-content/uploads/2017/01/wallpaper-image-nourriture-hd-13.jpg' },
      { name: 'Pizza-640', uri: require('./pizza-640.jpg') },
      { name: 'Coast', uri: 'https://thisismyhappiness.com/wp-content/uploads/2014/05/big-sur-bixby.jpg' },
      { name: 'Mountains', uri: 'https://www.highreshdwallpapers.com/wp-content/uploads/2011/09/Large-Format-HD-Wallpaper.jpg' },
      { name: 'Flowers', uri: 'https://media.ooreka.fr/public/image/plant/314/mainImage-source-11702050.jpg' },
      { name: 'Atx', uri: 'https://una.im/CSSgram/img/atx.jpg' },
      { name: 'Bike', uri: 'https://una.im/CSSgram/img/bike.jpg' },
      { name: 'Tahoe', uri: 'https://una.im/CSSgram/img/tahoe.jpg' },
      { name: 'Cacti', uri: 'https://una.im/CSSgram/img/cacti.jpg' },
      { name: 'LakeGeneva', uri: 'https://una.im/CSSgram/img/lakegeneva.jpg' }
    ],
    filters: [].concat.apply([], Array.from(Array(1)).map((x, i) => [
      { name: 'Normal', key: `Normal_${i}` },
      { name: 'Sepia', key: `Sepia_${i}` },
      { name: 'Saturate', key: `Saturate_${i}` },
      { name: 'HueRotate', key: `HueRotate_${i}` },
      { name: 'LuminanceToAlpha', key: `LuminanceToAlpha_${i}` },
      { name: 'Invert', key: `Invert_${i}` },
      { name: 'Grayscale', key: `Grayscale_${i}` },
      { name: 'Warm', key: `Warm_${i}` },
      { name: 'Cool', key: `Cool_${i}` },
      { name: 'Tint', key: `Tint_${i}` },
      { name: 'Night', key: `Night_${i}` },
      { name: 'Lsd', key: `Lsd_${i}` },
      { name: 'Browni', key: `Browni_${i}` },
      { name: 'Sharpen', key: `Sharpen_${i}` },
      { name: 'Emboss', key: `Emboss_${i}` },
      { name: 'EdgeDetection', key: `EdgeDetection_${i}` },
      { name: 'FuzzyGlass', key: `FuzzyGlass_${i}` },
      { name: '_1977', key: `_1977_${i}` }, { name: '_1977Compat', key: `_1977Compat_${i}` },
      { name: 'Aden', key: `Aden_${i}` }, { name: 'AdenCompat', key: `AdenCompat_${i}` },
      { name: 'Brannan', key: `Brannan_${i}` }, { name: 'BrannanCompat', key: `BrannanCompat_${i}` },
      { name: 'Brooklyn', key: `Brooklyn_${i}` }, { name: 'BrooklynCompat', key: `BrooklynCompat_${i}` },
      { name: 'Clarendon', key: `Clarendon_${i}` }, { name: 'ClarendonCompat', key: `ClarendonCompat_${i}` },
      { name: 'Earlybird', key: `Earlybird_${i}` }, { name: 'EarlybirdCompat', key: `EarlybirdCompat_${i}` },
      { name: 'Gingham', key: `Gingham_${i}` }, { name: 'GinghamCompat', key: `GinghamCompat_${i}` },
      { name: 'Hudson', key: `Hudson_${i}` }, { name: 'HudsonCompat', key: `HudsonCompat_${i}` },
      { name: 'Kelvin', key: `Kelvin_${i}` }, { name: 'KelvinCompat', key: `KelvinCompat_${i}` },
      { name: 'Lark', key: `Lark_${i}` }, { name: 'LarkCompat', key: `LarkCompat_${i}` },
      { name: 'Inkwell', key: `dInkwell_${i}` }, { name: 'InkwellCompat', key: `dInkwellCompat_${i}` },
      { name: 'Lofi', key: `Lofi_${i}` }, { name: 'LofiCompat', key: `LofiCompat_${i}` },
      { name: 'Maven', key: `Maven_${i}` }, { name: 'MavenCompat', key: `MavenCompat_${i}` },
      { name: 'Mayfair', key: `Mayfair_${i}` }, { name: 'MayfairCompat', key: `MayfairCompat_${i}` },
      { name: 'Moon', key: `Moon_${i}` }, { name: 'MoonCompat', key: `MoonCompat_${i}` },
      { name: 'Nashville', key: `Nashville_${i}` }, { name: 'NashvilleCompat', key: `NashvilleCompat_${i}` },
      { name: 'Perpetua', key: `Perpetua_${i}` }, { name: 'PerpetuaCompat', key: `PerpetuaCompat_${i}` },
      { name: 'Reyes', key: `Reyes_${i}` }, { name: 'ReyesCompat', key: `ReyesCompat_${i}` },
      { name: 'Rise', key: `Rise_${i}` }, { name: 'RiseCompat', key: `RiseCompat_${i}` },
      { name: 'Slumber', key: `Slumber_${i}` }, { name: 'SlumberCompat', key: `SlumberCompat_${i}` },
      { name: 'Stinson', key: `Stinson_${i}` }, { name: 'StinsonCompat', key: `StinsonCompat_${i}` },
      { name: 'Toaster', key: `Toaster_${i}` }, { name: 'ToasterCompat', key: `ToasterCompat_${i}` },
      { name: 'Valencia', key: `Valencia_${i}` }, { name: 'ValenciaCompat', key: `ValenciaCompat_${i}` },
      { name: 'Walden', key: `Walden_${i}` }, { name: 'WaldenCompat', key: `WaldenCompat_${i}` },
      { name: 'Willow', key: `Willow_${i}` }, { name: 'WillowCompat', key: `WillowCompat_${i}` },
      { name: 'Xpro2', key: `Xpro2_${i}` }, { name: 'Xpro2Compat', key: `Xpro2Compat_${i}` }
    ]))
  };

  componentDidMount() {
    // this.interval = setInterval(() => this.setState({ t: Date.now() }), 7000);
  }

  componentWillUnmount() {
    // clearInterval(this.interval);
  }

  renderList() {
    return (
      <FlatList
        contentContainerStyle={styles.container}
        data={this.state.filters}
        renderItem={this.renderFilter}
      />
    );
  }


  renderSelect() {
    const { selectedFilter, first, filters, images, selectedImage } = this.state;

    return (
      <ScrollView contentContainerStyle={styles.container}>
        <Pick
          style={styles.picker}
          selectedValue={selectedFilter}
          onValueChange={this.filterChanged}
        >
          {filters.map(({ name, key }) => <PickItem value={key} label={name} key={key} />)}
        </Pick>
        <Pick
          style={styles.picker}
          selectedValue={selectedImage}
          onValueChange={this.valueChanged}
        >
          {images.map(({ name }) => <PickItem value={name} label={name} key={name} />)}
        </Pick>
        <CSSGramItem
          filter={filters.find(({ key }) => key === selectedFilter).name}
          image={images.find(({ name }) => name === selectedImage).uri}
        />
      </ScrollView>
    );
  }

  filterChanged = (item) => {
    this.setState({ selectedFilter: item })
  }

  valueChanged = (item) => {
    this.setState({ selectedImage: item })
  }

  renderBlends() {
    return (
      <View>
        <Text>'TEST'</Text>
      </View>
    );
    // const { selectedBlend, blends, firstSelectedImage, secondSelectedImage, blendImages } = this.state

    // return (
    //   <View>
    //     <Pick
    //       style={styles.picker}
    //       selectedValue={selectedBlend}
    //       onValueChange={this.blendChanged}
    //     >
    //       {blends.map((blend) => <PickItem value={blend} label={blend} key={blend} />)}
    //     </Pick>
    //     <Pick
    //       style={styles.picker}
    //       selectedValue={firstSelectedImage}
    //       onValueChange={this.firstImageChanged}
    //     >
    //       {blendImages.map(({ name }) => <PickItem value={name} label={name} key={name} />)}
    //     </Pick>
    //     <Pick
    //       style={styles.picker}
    //       selectedValue={secondSelectedImage}
    //       onValueChange={this.secondImageChanged}
    //     >
    //       {blendImages.map(({ name }) => <PickItem value={name} label={name} key={name} />)}
    //     </Pick>
    //     {this.renderBlend(
    //       selectedBlend,
    //       blendImages.find(b => b.name === firstSelectedImage).item,
    //       blendImages.find(b => b.name === secondSelectedImage).item,
    //     )}
    //   </View>
    // );
  }

  renderBlend(name, firstSelectedImage, secondSelectedImage) {
    const blend = `${name}Blend`
    const blendColor = `${blend}Color`

    const config = {
      name: blend,
      dstResizeMode: 'CONTAIN',
      srcResizeMode: 'CONTAIN',
      // dstAnchor: { x: 0, y: 1 },
      // srcPosition: { x: 0, y: 1 },'
      // srcAnchor: { x: 0, y: 1 },
      // srcPosition: { x: 0, y: 1 }
    }

    return (
      <View>
        <Text>{blend}</Text>
        <ImageFilter
          key={blend + '1'}
          config={{
            dstImage: firstSelectedImage,
            srcImage: secondSelectedImage,
            ...config
          }}
        />
        <ImageFilter
          key={blend + '2'}
          config={{
            dstImage: secondSelectedImage,
            srcImage: firstSelectedImage,
            ...config
          }}
        />
        <Text>{blendColor}</Text>
        <ImageFilter
          key={blendColor}
          config={{
            name: blendColor,
            dstImage: firstSelectedImage,
            srcColor: '#0080ff80',
          }}
        />
      </View>
    )
  }

  blendChanged = (selectedBlend) => {
    this.setState({ selectedBlend })
  }

  firstImageChanged = (firstSelectedImage) => {
    this.setState({ firstSelectedImage })
  }

  secondImageChanged = (secondSelectedImage) => {
    this.setState({ secondSelectedImage })
  }

  image(name) {
    return this.state.blendImages.find(x => x.name === name).item
  }

  render() {
    return (
      <ScrollView contentContainerStyle={{ alignItems: 'center' }}>
        <Switch
          value={this.state.showList}
          onValueChange={(showList) => this.setState({ showList })}
        />
        {this.state.showList ? this.renderBlends() : this.renderSelect()}
      </ScrollView>
    )
  }

  renderFilter = ({ item }) => (
    item.uri
      ? <Image
          style={{ width: 360, height: 360, borderRadius: 25 }}
          source={item}
          blurRadius={5}
          resizeMode={'contain'}
        />
      : <CSSGramItem filter={item.name} />
  )
}

const styles = StyleSheet.create({
  filter: {
  },
  textWrap: {
    position: 'absolute',
    right: 10,
    bottom: 10,
    paddingHorizontal: 5,
    paddingVertical: 3,
    borderRadius: 5,
    backgroundColor: 'rgba(0, 0, 0, 0.35)'
  },
  text: {
    fontSize: 20,
    fontWeight: 'bold',
    color: 'white'
  },
  container: {
    width: '100%',
    alignItems: 'center'
  },
  welcome: {
    fontSize: 20,
    textAlign: 'center',
    margin: 10,
  },
  instructions: {
    textAlign: 'center',
    color: '#333333',
    marginBottom: 5,
  },
  picker: {
    alignSelf: 'stretch',
    maxHeight: 150
  },
  image: imageStyle,
  halfImage: { width: 180, height: 180 }
});
